### PR TITLE
Fix/when retry request

### PR DESCRIPTION
**Pull request by agranado2k**
- [x] Tests written and linted [ℹ︎](https://github.com/snyk/general/wiki/Tests)
- [ ] Documentation written [ℹ︎](https://github.com/snyk/general/wiki/Documentation)
- [x] Commit history is tidy [ℹ︎](https://github.com/snyk/general/wiki/Git)

### What this does

It has a small refactoring moving the condition that decide if k8s-monitor should or not retry a request. Also it includes a new error code that allows k8s-monitor decides if it should retry a request or not.

### Notes for the reviewer

This new error code was introduced because sometimes kubernetes cluster can have timeout caused by slow DNS lookup. And k8s-monitor should retry the request if it stills having remaining retries.

### More information

- [Jira ticket RUN-882](https://snyksec.atlassian.net/browse/RUN-882)

### CODE REVIEW

File: src/transmitter/index.ts [hunk 1]
```diff
@@ -114,3 +112,15 @@ async function retryRequest(verb: NeedleHttpVerbs, url: string, payload: object)
 
   return {response, attempt};
 }
+
+function shouldRetryRequest(err: IRequestError, networkErrorMessages: string[], stillHaveRetries: boolean): boolean {
+  if (!stillHaveRetries) {
```
Comment (shaimendel) [inline src/transmitter/index.ts:117]: that's not a real review, just one comment about this function :) I don't like `else` when there is no need. just remove all the `else`:
```
if (something) {
  return something;
}

if (something else) {
  return something else;
}

return false;
```

Comment (agranado2k) [reply src/transmitter/index.ts:117]: Thanks @shaimendel! Definitely it's a good a good comment to get the code style!

File: src/transmitter/index.ts [hunk 2]
```diff
@@ -114,3 +112,15 @@ async function retryRequest(verb: NeedleHttpVerbs, url: string, payload: object)
 
   return {response, attempt};
 }
+
+function shouldRetryRequest(err: IRequestError, networkErrorMessages: string[], stillHaveRetries: boolean): boolean {
```
Comment (ivanstanev) [inline src/transmitter/index.ts:116]: I think that `networkErrorMessages` is really only used within this `shouldRetryRequest()` function so it probably doesn't need to be passed as a variable any more. I think also currently the array gets allocated every time the function `retryRequest()` is called, so we probably don't need to do that! How about moving the `networkErrorMessages[]` as a const global variable in the current module?

File: src/transmitter/index.ts [hunk 3]
```diff
@@ -114,3 +112,15 @@ async function retryRequest(verb: NeedleHttpVerbs, url: string, payload: object)
...
+  if (!stillHaveRetries) {
+    return false;
+  }
+
+  if (err.code === RequestCodeError.ECONNRESET && networkErrorMessages.includes(err.message)) {
```
Comment (ivanstanev) [inline src/transmitter/index.ts:121]: ➕ 1️⃣  on the Enum idea, I also think it's a really nice feature to use! However, in this current context I think it doesn't introduce benefits. It gives an extra layer of indirection (maybe you want to check what's the value of the enum?) and is a bit longer to type 😄 What do you think of just keeping it like before `(err.code === 'ECONNRESET')` as it is very explicit and easy to read?

File: test/system/kind.test.ts [hunk 4]
```diff
@@ -135,7 +135,7 @@ tap.test('Kubernetes-Monitor with KinD', async (t) => {
   require('../../src');
 
   // TODO: replace with being event driven?
-  // will still need SOME timeout 
+  // will still need SOME timeout
```
Comment (ivanstanev) [inline test/system/kind.test.ts:146]: This change probably shouldn't be part of this commit. Maybe committed by accident since your linting tool has fixed it?
While it does remove an extra space, it's in a comment, so really not that relevant and it doesn't improve the quality of the codebase 😆

File: src/transmitter/types.ts [hunk 5]
```diff
@@ -72,3 +72,12 @@ export interface IResponseWithAttempts {
   response: NeedleResponse;
   attempt: number;
 }
+
+export interface IRequestError {
+  code: string,
+  message: string
```
Comment (Shesekino) [inline src/transmitter/types.ts:78]: comma at the end of the line please

File: src/transmitter/types.ts [hunk 6]
```diff
@@ -72,3 +72,12 @@ export interface IResponseWithAttempts {
...
+}
+
+export enum RequestCodeError {
+  ECONNRESET = 'ECONNRESET'
```
Comment (Shesekino) [inline src/transmitter/types.ts]: comma at the end of the line I think?

File: src/transmitter/index.ts [hunk 7]
```diff
@@ -122,5 +122,9 @@ function shouldRetryRequest(err: IRequestError, networkErrorMessages: string[],
     return true;
   }
 
+  if (err.code === RequestCodeError.EAI_AGAIN) {
```
Comment (Shesekino) [inline src/transmitter/index.ts:125]: no need to specify "k8s-monitor" in the commit title; this is the Kubernetes-Monitor repo. unless specified otherwise, everything here is in the context of the kubernetes monitor.
so in this case, ```fix: retry transmitting data on EAI_AGAIN error codes``` might be more fitting

File: src/transmitter/types.ts [hunk 8]
```diff
@@ -79,5 +79,6 @@ export interface IRequestError {
 }
 
 export enum RequestCodeError {
-  ECONNRESET = 'ECONNRESET'
+  ECONNRESET = 'ECONNRESET',
+  EAI_AGAIN = 'EAI_AGAIN'
```
Comment (Shesekino) [inline src/transmitter/types.ts:83]: comma at the end of the line

File: test/system/kind.test.ts [hunk 9]
```diff
@@ -103,6 +103,14 @@ tap.test('Kubernetes-Monitor with KinD', async (t) => {
       message: 'socket hang up',
     });
 
+  nock('https://kubernetes-upstream.snyk.io')
```
Comment (Shesekino) [inline test/system/kind.test.ts:106]: nice!

File: src/transmitter/types.ts (hunk 5 above)
Comment (agatakrajewska) [reply src/transmitter/types.ts:78]: I think it should be colons, like: 
```
export interface IRequestError {
  code: string;
  message: string;
}
```

**Comment by agatakrajewska** [comment]
Yay, well done! 🙂 🥳 
Great approach with separate commits introducing refactoring first 👍
My main comment would also be about the commit messages, as @ivanstanev explained in detail, it's important to see why we are making certain changes, rather than pointing them out 🙂And also using just text without bullets and with punctuation +1️⃣from me as well, makes it super easy to read.
